LBAC-2 Added implementation to add the locations to the users on registration

suthagar23 opened the pull request:
# Description

This PR contains the implementation to assign the location to the users on registration. It contains two custom fragments for,

- Add/Edit the location in the user dashboard
- View the location information in the user dashboard

# Ticket 
Ticket : https://issues.openmrs.org/browse/LBAC-2

dkayiwa commented on api/src/main/java/org/openmrs/module/locationbasedaccess/LocationBasedAccessConstants.java:
```diff
@@ -13,4 +13,5 @@
 public class LocationBasedAccessConstants {
 
     public static final  String LOCATION_ATTRIBUTE_GLOBAL_PROPERTY_NAME = "locationbasedaccess.locationAttributeUuid";
+    public static final  String USER_PROPERTY_LOCATION_ATTRIBUTE_NAME = "locationUuid";
```
Is the above constant name really correct?

suthagar23 replied:
changed to "LOCATION_USER_PROPERTY_NAME"

dkayiwa commented on api/src/main/resources/messages.properties:
```diff
@@ -1,3 +1,4 @@
 ${project.parent.artifactId}.title=Location based access control module
 
 locationbasedaccess.patientLocation=Patient Location
```
Are we still using locationbasedaccess.patientLocation for anything?

suthagar23 replied:
For showing the label on the patient registration dashboard.

dkayiwa commented on omod/src/main/resources/apps/createUser_extension.json:
```diff
@@ -0,0 +1,26 @@
+      "type": "userProperty",
+      "userPropertyName" : "locationId",
+      "label": "Select Location",
+      "uuid": "8b5c95ef-103c-41bc-9f24-368b8f77e070",
```
What is the above uuid for?

suthagar23 replied:
removed

dkayiwa commented on omod/src/main/resources/apps/createUser_extension.json:
```diff
@@ -0,0 +1,26 @@
+      "title" : "Location",
+      "type": "userProperty",
+      "userPropertyName" : "locationId",
+      "label": "Select Location",
```
Shouldn't we localise these strings?

suthagar23 replied:
Yes, we need to do it. Since other extensions also need to have it, Can we do these all with a different ticket?

dkayiwa replied:
I thought you would just replace them with message codes hence not requiring a separate ticket.

dkayiwa commented on omod/src/main/java/org/openmrs/module/locationbasedaccess/fragment/controller/manageAccount/ViewUserLocationFragmentController.java:
```diff
@@ -0,0 +1,28 @@
+public class ViewUserLocationFragmentController {
+
+    public void controller(FragmentModel model, FragmentConfiguration config) {
+        Object userId = config.getAttribute("userId");
```
Did you intentionally decide to use the integer userId instead of user uuid?

suthagar23 replied:
Just updated

dkayiwa commented on omod/src/main/java/org/openmrs/module/locationbasedaccess/fragment/controller/manageAccount/ViewUserLocationFragmentController.java:
```diff
@@ -0,0 +1,28 @@
+        if (userId != null) {
+            User user = Context.getUserService().getUser(Integer.parseInt(userId.toString()));
+            if (user != null) {
+                String locationProperty = user.getUserProperty("locationId");
```
Shouldn't the property be locationUuid?

suthagar23 replied:
Just updated with constants

dkayiwa commented on omod/src/main/java/org/openmrs/module/locationbasedaccess/fragment/controller/manageAccount/AddLocationFragmentController.java:
```diff
@@ -0,0 +1,46 @@
+        }
+        else if (person != null) {
+            Location location = LocationUtils.getPersonLocation(person);
+            if(location!=null) {
```
The formatting would always be better if you take advantage of spaces. For instance the above would be like:
if (location != null) {

dkayiwa commented on omod/src/main/java/org/openmrs/module/locationbasedaccess/fragment/controller/field/LocationsFragmentController.java:
```diff
@@ -29,14 +27,9 @@ public void controller(FragmentModel model,
-                    model.addAttribute("selectedLocationUuid", patientLocation.getUuid());
-                }
+            Location location = LocationUtils.getPersonLocation(patient.getPerson());
+            if(location!=null) {
```
Same issue with the above line formatting.

dkayiwa commented on omod/src/main/java/org/openmrs/module/locationbasedaccess/fragment/controller/manageAccount/AddLocationFragmentController.java:
```diff
@@ -0,0 +1,46 @@
+
+        if (user != null) {
+            Object userProperty = user.getUserProperty(LocationBasedAccessConstants.LOCATION_USER_PROPERTY_NAME);
+            if (userProperty!=null) {
```
Can you also put spaces for the line above?

suthagar23 replied:
Added Thanks

dkayiwa commented on omod/src/main/java/org/openmrs/module/locationbasedaccess/fragment/controller/manageAccount/ViewUserLocationFragmentController.java:
```diff
@@ -0,0 +1,29 @@
+public class ViewUserLocationFragmentController {
+
+    public void controller(FragmentModel model, FragmentConfiguration config) {
+        Object userUuid = config.getAttribute("userUuid");
```
Can we name it as userId instead of userUuid?

suthagar23 replied:
Changes to userId.

dkayiwa commented on omod/src/main/resources/apps/createUser_extension.json:
```diff
@@ -0,0 +1,25 @@
+    "id": "${project.parent.groupId}.${project.parent.artifactId}.createUser.customFragments.userViewLocation",
+    "extensionPointId": "createUser.manageAccountUserViewFragments",
+    "extensionParams": {
+      "label": "locationbasedaccess.location",
```
After replacing the labels, did you test to confirm that all is well?

dkayiwa commented on omod/src/main/webapp/fragments/field/locations.gsp:
```diff
@@ -3,7 +3,7 @@
     <div>
         <p class="left">
             <label>
-                Please select the location
```
Can we drop the "Please select the"? and just use Location?

dkayiwa replied:
The same applies else where you use please select the location instead of just Location

suthagar23 replied:
@dkayiwa  I have changed that to location.

dkayiwa commented:
Did you also change the extension ids?

suthagar23 commented:
Yes @dkayiwa  I have changed and I have updated the docs also - https://wiki.openmrs.org/display/docs/Adding+custom+fragments+to+Manage+Account+dashboard

dkayiwa commented:
In the docs, i still see something like createUser.customFragments.userLocation

dkayiwa commented on omod/src/main/resources/apps/createUser_extension.json:
```diff
@@ -0,0 +1,22 @@
+[
+  {
+    "id": "${project.parent.groupId}.${project.parent.artifactId}.createUser.customFragments.userLocation",
```
I thought we changed from "createUser"?

suthagar23 replied:
Just updated @dkayiwa  
Sorry, I thought it was just ids specifically for this module.

dkayiwa commented on omod/src/main/resources/apps/createUser_extension.json:
```diff
@@ -0,0 +1,22 @@
+    }
+  },
+  {
+    "id": "${project.parent.groupId}.${project.parent.artifactId}.userAccount.userViewLocation",
```
The above lack "Property"

suthagar23 replied:
Just added @dkayiwa

dkayiwa commented on omod/src/main/resources/apps/createUser_extension.json:
```diff
@@ -0,0 +1,22 @@
+[
+  {
+    "id": "${project.parent.groupId}.${project.parent.artifactId}.userAccount.userEditLocation",
```
The above lacks "Property"

suthagar23 replied:
just added
